### PR TITLE
PluginManager: Make Plugins, Renderer and DataSources non-global

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
Make `Plugins`, `Renderer` and `DataSources` in pkg/plugins/manager non-global, i.e. accessible via the `plugins.Manager` service. The plugin manager is also abstracted behind an interface (`plugins.Manager`), to break dependency cycles.

**Which issue(s) this PR fixes**:

Fixes part of #29022.

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
This is looking really good!

**Comment by aknuds1** [comment]
Having to fix integration tests, that turns out to be easier said than done!

**Comment by aknuds1** [comment]
Fixed integration test problems.

**Comment by aknuds1** [comment]
Gah, MySQL integration tests fail although Postgres don't :(

**Review by torkelo — APPROVED** [review]
Great job on this one!

**Comment by marefr** [comment]
I did try reproduce mysql integration tests locally, but couldn't. I only ran the services/dashboard package tests though. Wonder whether running all of them somehow interfere/affect? @aknuds1 have you found any clues?

File: pkg/infra/usagestats/usage_stats.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (uss *UsageStatsService) GetUsageReport(ctx context.Context) (UsageReport,
 	metrics["stats.playlist.count"] = statsQuery.Result.Playlists
 	metrics["stats.plugins.apps.count"] = len(manager.Apps)
 	metrics["stats.plugins.panels.count"] = len(manager.Panels)
-	metrics["stats.plugins.datasources.count"] = len(manager.DataSources)
+	metrics["stats.plugins.datasources.count"] = uss.PluginManager.NumDataSources()
```
Comment (marefr) [inline pkg/infra/usagestats/usage_stats.go:61]:
```suggestion
	metrics["stats.plugins.datasources.count"] = uss.PluginManager.DataSourceCount()
```

File: pkg/infra/usagestats/usage_stats_test.go [hunk 2]
```diff
@@ -271,7 +271,7 @@ func TestMetrics(t *testing.T) {
 			assert.Equal(t, getSystemStatsQuery.Result.Playlists, metrics.Get("stats.playlist.count").MustInt64())
 			assert.Equal(t, len(manager.Apps), metrics.Get("stats.plugins.apps.count").MustInt())
 			assert.Equal(t, len(manager.Panels), metrics.Get("stats.plugins.panels.count").MustInt())
-			assert.Equal(t, len(manager.DataSources), metrics.Get("stats.plugins.datasources.count").MustInt())
+			assert.Equal(t, uss.PluginManager.NumDataSources(), metrics.Get("stats.plugins.datasources.count").MustInt())
```
Comment (marefr) [inline pkg/infra/usagestats/usage_stats_test.go:274]:
```suggestion
			assert.Equal(t, uss.PluginManager.DataSourceCount(), metrics.Get("stats.plugins.datasources.count").MustInt())
```

File: pkg/infra/usagestats/usage_stats_test.go [hunk 3]
```diff
@@ -508,37 +508,51 @@ func (aum *alertingUsageMock) QueryUsageStats() (*alerting.UsageStats, error) {
 	}, nil
 }
 
+type fakePluginManager struct {
+	manager.PluginManager
+
+	dataSources map[string]*plugins.DataSourcePlugin
+}
+
+func (pm fakePluginManager) NumDataSources() int {
```
Comment (marefr) [inline pkg/infra/usagestats/usage_stats_test.go:517]:
```suggestion
func (pm fakePluginManager) DataSourceCount() int {
```

File: pkg/plugins/ifaces.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	// Renderer gets the renderer plugin.
+	Renderer() *RendererPlugin
+	// GetDataSource gets a data source plugin with a certain ID.
+	GetDataSource(id string) *DataSourcePlugin
+	// GetDataPlugin gets a data plugin with a certain ID.
+	GetDataPlugin(id string) DataPlugin
+	// GetPlugin gets a plugin with a certain ID.
+	GetPlugin(id string) *PluginBase
+	// NumDataSources gets the number of data sources.
+	NumDataSources() int
```
Comment (marefr) [inline pkg/plugins/ifaces.go:21]:
```suggestion
	// DataSourceCount gets the count of data sources.
	DataSourceCount() int
```

File: pkg/plugins/manager/manager.go [hunk 5]
```diff
@@ -178,6 +182,48 @@ func (pm *PluginManager) Run(ctx context.Context) error {
...
+func (pm *PluginManager) DataSources() []*plugins.DataSourcePlugin {
+	var rslt []*plugins.DataSourcePlugin
+	for _, ds := range pm.dataSources {
+		rslt = append(rslt, ds)
+	}
+
+	return rslt
+}
+
+func (pm *PluginManager) NumDataSources() int {
```
Comment (marefr) [inline pkg/plugins/manager/manager.go:202]:
```suggestion
func (pm *PluginManager) DataSourceCount() int {
```

File: pkg/api/dashboard.go [hunk 6]
```diff
@@ -122,7 +121,8 @@ func (hs *HTTPServer) GetDashboard(c *models.ReqContext) response.Response {
 		meta.FolderUrl = query.Result.GetUrl()
 	}
 
-	provisioningData, err := dashboards.NewProvisioningService().GetProvisionedDashboardDataByDashboardID(dash.Id)
+	svc := dashboards.NewService(hs.SQLStore, hs.SQLStore, hs.DataService)
```
Comment (marefr) [inline pkg/api/dashboard.go:124]: Nit, providing sqlstore as both first and second argument looks a bit strange. But inspecting the function signature it makes sense.

**Review by marefr — COMMENTED** [review]
Great work with this! Left some suggestions/comments.

File: pkg/api/dashboard.go (hunk 6 above)
Comment (aknuds1) [reply pkg/api/dashboard.go:124]: @marefr Yeah, it has to be that way when using two different interfaces, the alternative would be to join into one.

**Comment by aknuds1** [comment]
> I did try reproduce mysql integration tests locally, but couldn't. I only ran the services/dashboard package tests though. Wonder whether running all of them somehow interfere/affect? @aknuds1 have you found any clues?

@marefr sorry, been all caught up in meetings since.

File: pkg/api/dashboard.go (hunk 6 above)
Comment (marefr) [reply pkg/api/dashboard.go]: It's fine as is.

**Comment by aknuds1** [comment]
@marefr I rewrote the dashboard service integration tests (to use Go stdlib) and made sure to remove usages of the global Xorm engine (I'm not aware of any more usages in relevant paths). It looks as if MySQL integration tests still fail on the first attempt, but go through on the second one (the CI pipeline retries twice I think).

**Review by torkelo — APPROVED** [review]
Awesome job! Merge when green :)

**Pull request merged**